lib/manageiq/providers/workflows/runner.rb

```diff
@@ -0,0 +1,86 @@
+module ManageIQ
+  module Providers
+    module Workflows
+      class Runner
+        include Vmdb::Logging
+
+        class << self
+          def runner
+            @runner ||= new.tap(&:start)
+          end
+        end
+
+        attr_reader :workflows
+
+        def initialize
+          require "floe"
+          require "concurrent/hash"
+
+          @workflows          = Concurrent::Hash.new
+          @docker_wait_thread = nil
+        end
+
+        def start
+          $workflows_log.debug("Runner: Starting workflows runner...")
+          self.docker_wait_thread = Thread.new { docker_wait }
+          $workflows_log.debug("Runner: Starting workflows runner...Complete")
+        end
+
+        def stop
+          $workflows_log.debug("Runner: Stopping workflows runner...")
+          stop_thread(docker_wait_thread)
+
+          self.docker_wait_thread = nil
+          $workflows_log.debug("Runner: Stopping workflows runner...Complete")
+        end
+
+        def add_workflow(workflow, queue_args)
+          workflows[workflow.id] ||= [workflow, queue_args]
+        end
+
+        def delete_workflow(workflow)
+          workflows.delete(workflow.id)
+        end
+
+        private
+
+        attr_accessor :docker_wait_thread
+
+        def docker_wait
+          loop do
```

Comment:
does this loop handle an issue with docker

e.g.: in development we restart our docker process.
that will probably kick us out of `docker_runner.wait`

then the loop will swoop back, connect, and continue?

Reply:
So both docker :tm: and podman use a command line utility to get events with a timeout, and kubernetes uses watches.  All three use a loop internally to retry on failure / break out if timeout has been exceeded.